vms/platformvm/txs/mempool/mempool.go

```diff
@@ -5,42 +5,291 @@ package mempool
+	weights gas.Dimensions,
+	gasCapacity gas.Gas,
+	avaxAssetID ids.ID,
+	registerer prometheus.Registerer,
+) (*Mempool, error) {
+	numTxsMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "count",
+		Help:      "number of transactions in the mempool",
+	})
+
+	gasAvailableMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "gas_available",
+		Help:      "amount of remaining gas available",
+	})
+
+	if err := errors.Join(
+		registerer.Register(numTxsMetric),
+		registerer.Register(gasAvailableMetric),
+	); err != nil {
+		return nil, fmt.Errorf("failed to register metrics: %w", err)
 	}
-	pool := txmempool.New[*txs.Tx](
-		metrics,
-	)
-	return &Mempool{Mempool: pool}, nil
+
+	return &Mempool{
+		weights:     weights,
+		gasCapacity: gasCapacity,
+		avaxAssetID: avaxAssetID,
+		maxHeap: heap.NewMap[ids.ID, heapTx](func(a, b heapTx) bool {
+			return a.GasPrice > b.GasPrice
+		}),
+		minHeap: heap.NewMap[ids.ID, heapTx](func(a, b heapTx) bool {
+			return a.GasPrice < b.GasPrice
+		}),
+		consumedUTXOs:      setmap.New[ids.ID, ids.ID](),
+		droppedTxIDs:       lru.NewCache[ids.ID, error](64),
+		numTxsMetric:       numTxsMetric,
+		gasAvailableMetric: gasAvailableMetric,
+	}, nil
 }
 
 func (m *Mempool) Add(tx *txs.Tx) error {
-	switch tx.Unsigned.(type) {
-	case *txs.AdvanceTimeTx:
-		return ErrCantIssueAdvanceTimeTx
-	case *txs.RewardValidatorTx:
-		return ErrCantIssueRewardValidatorTx
-	default:
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	ins, outs, producedAVAX, err := utxo.GetInputOutputs(tx.Unsigned)
+	if err != nil {
+		return fmt.Errorf("failed to get utxos: %w", err)
+	}
+
+	consumedAVAX := uint64(0)
+	for _, utxo := range ins {
+		if utxo.AssetID() != m.avaxAssetID {
+			continue
+		}
+
+		consumedAVAX += utxo.In.Amount()
+	}
+
+	if consumedAVAX == 0 {
+		return errMissingConsumedAVAX
+	}
+
+	for _, utxo := range outs {
+		if utxo.AssetID() != m.avaxAssetID {
+			continue
+		}
+
+		producedAVAX += utxo.Out.Amount()
+	}
+
+	_, gasUsed, err := m.meter(tx.Unsigned)
+	if err != nil {
+		return fmt.Errorf("failed to meter tx: %w", err)
+	}
+
+	heapTx := heapTx{
+		Tx:       tx,
+		gasUsed:  gasUsed,
+		GasPrice: float64(consumedAVAX-producedAVAX) / float64(gasUsed),
+	}
+
+	// Try to evict a lower paying tx if we do not have enough remaining gas
+	// capacity
+	if next := m.currentGas + gasUsed; next > m.gasCapacity {
+		if err := m.tryEvictTx(next, heapTx); err != nil {
+			return err
+		}
+	}
+
+	if m.consumedUTXOs.HasOverlap(tx.InputIDs()) {
+		return mempool.ErrConflictsWithOtherTx
+	}
+
+	m.maxHeap.Push(tx.TxID, heapTx)
+	m.minHeap.Push(tx.TxID, heapTx)
+	m.consumedUTXOs.Put(tx.TxID, tx.InputIDs())
+	m.droppedTxIDs.Evict(tx.TxID)
+	m.currentGas += gasUsed
+
+	m.updateMetrics()
+
+	return nil
+}
+
+// Evict a tx if it is lower paying and the resulting total gas is within
+// capacity
+func (m *Mempool) tryEvictTx(next gas.Gas, tx heapTx) error {
```

Comment:
This method feels a bit unsatisfying to me. If the lowest paying tx is really small, then it may never be able to get evicted (which seems unusual to me). On the flip side, evicting multiple transactions gets a bit complex as well. For example, we may need to evict two txs. Based on the gas price we evict the first, but the second shouldn't be evicted based on the gas price.

I suppose we could "re-add" any txs that we have evicted if the eviction failed (which is kind of what we do for checking for conflicts currently)

Thoughts?

Reply:
As discussed offline I can get this to just iterate over the heap in the worst-case and just re-add txs if eviction fails